<a href="https://colab.research.google.com/github/farizalmustaqim/Proyek-Sentiment-Analysis-NLP/blob/main/Test_Pra_Interview_DA_Botika.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# __Proyek Sentiment Analysis NLP__
Nama: Fariz Al Mustaqim


# __1. Pelabelan Dataset__

Pelabelan dataset teks menggunakan sentimen analisis dari packages transformers

In [ ]:
!pip install transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 4.4 MB 5.2 MB/s 
     |████████████████████████████████| 101 kB 9.9 MB/s 
     |████████████████████████████████| 596 kB 58.3 MB/s 
     |████████████████████████████████| 6.6 MB 38.6 MB/s 
  Attempting uninstall: pyyaml
    Found existing installation: PyYAML 3.13
    Uninstalling PyYAML-3.13:
      Successfully uninstalled PyYAML-3.13


In [ ]:
from transformers import AutoTokenizer, AutoModel
tokenizer = AutoTokenizer.from_pretrained("indolem/indobertweet-base-uncased")
model = AutoModel.from_pretrained("indolem/indobertweet-base-uncased")

Downloading:   0%|          | 0.00/2.00 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/1.08k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/230k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/2.00 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/112 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/424M [00:00<?, ?B/s]

Some weights of the model checkpoint at indolem/indobertweet-base-uncased were not used when initializing BertModel: ['cls.predictions.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.dense.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.decoder.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.decoder.weight']
- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


In [ ]:
model

BertModel(
  (embeddings): BertEmbeddings(
    (word_embeddings): Embedding(31923, 768, padding_idx=0)
    (position_embeddings): Embedding(512, 768)
    (token_type_embeddings): Embedding(2, 768)
    (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (encoder): BertEncoder(
    (layer): ModuleList(
      (0): BertLayer(
        (attention): BertAttention(
          (self): BertSelfAttention(
            (query): Linear(in_features=768, out_features=768, bias=True)
            (key): Linear(in_features=768, out_features=768, bias=True)
            (value): Linear(in_features=768, out_features=768, bias=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (output): BertSelfOutput(
            (dense): Linear(in_features=768, out_features=768, bias=True)
            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
            (dropout): Dropout(p=0.1, inplace=False)
          

In [ ]:
from transformers import pipeline

pretrained_name = "sahri/indonesiasentiment"

nlp = pipeline(
    "sentiment-analysis",
    model=pretrained_name,
    tokenizer=pretrained_name
)

nlp("Saya udah gak ngerti.")

Downloading:   0%|          | 0.00/932 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/476M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/328 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/789k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/505k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/1.31M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/239 [00:00<?, ?B/s]

[{'label': 'negative', 'score': 0.9990731477737427}]

In [ ]:
nlp("Makasih")

[{'label': 'positive', 'score': 0.9978877902030945}]

In [ ]:
nlp("jelek")

[{'label': 'negative', 'score': 0.9985276460647583}]

In [ ]:
nlp("rakyatnya ngantri minyak goreng")

[{'label': 'neutral', 'score': 0.9958090782165527}]

**Mengimport Dataset** 

In [ ]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/farizalmustaqim/Test-Pra-Interview-DA-Botika/main/Data%20Chat%20User%20ke%20Bot%20(1).csv', on_bad_lines='skip')
df

data teks
0                             Makasii ????
1                  Sekarang jam berapa ya?
2     Hp gua udah 4g gausah diaktifin lagi
3                                 Semlekom
4    Sejag kapan nama saiia blue (Har Har)
..                                     ...
378      Punya anjir. Bacod mulu dari tadi
379          Terima kasih atas infonya bak
380                Can you speak English ?
381                         Namanya siapa?
382                   Mana gua tau selesai

[383 rows x 1 columns]

In [ ]:
#Memasukan label pada Dataset
df1 = nlp(df['data teks'].tolist())
df1


[{'label': 'negative', 'score': 0.994078516960144},
 {'label': 'negative', 'score': 0.9949683547019958},
 {'label': 'negative', 'score': 0.9980512857437134},
 {'label': 'negative', 'score': 0.9917022585868835},
 {'label': 'neutral', 'score': 0.9896272420883179},
 {'label': 'negative', 'score': 0.9983376264572144},
 {'label': 'neutral', 'score': 0.9958940744400024},
 {'label': 'negative', 'score': 0.9850267767906189},
 {'label': 'positive', 'score': 0.8499382734298706},
 {'label': 'neutral', 'score': 0.8948452472686768},
 {'label': 'positive', 'score': 0.9984021782875061},
 {'label': 'positive', 'score': 0.6948410868644714},
 {'label': 'positive', 'score': 0.9982927441596985},
 {'label': 'neutral', 'score': 0.9972150325775146},
 {'label': 'positive', 'score': 0.8141495585441589},
 {'label': 'positive', 'score': 0.9989284873008728},
 {'label': 'positive', 'score': 0.750587522983551},
 {'label': 'positive', 'score': 0.6069244146347046},
 {'label': 'negative', 'score': 0.6270015239715576},

In [ ]:
# Membuat pandas Dataframe untuk fitur label dan score
df2= pd.DataFrame(df1)
df2

label     score
0    negative  0.994079
1    negative  0.994968
2    negative  0.998051
3    negative  0.991702
4     neutral  0.989627
..        ...       ...
378  negative  0.944670
379  positive  0.965316
380  negative  0.997171
381  negative  0.893093
382  negative  0.997722

[383 rows x 2 columns]

In [ ]:
# Menggabungkan Dataframe fitur/column data teks, label dan score
df3 = pd.concat([df,df2],axis=1)
df3

data teks     label     score
0                             Makasii ????  negative  0.994079
1                  Sekarang jam berapa ya?  negative  0.994968
2     Hp gua udah 4g gausah diaktifin lagi  negative  0.998051
3                                 Semlekom  negative  0.991702
4    Sejag kapan nama saiia blue (Har Har)   neutral  0.989627
..                                     ...       ...       ...
378      Punya anjir. Bacod mulu dari tadi  negative  0.944670
379          Terima kasih atas infonya bak  positive  0.965316
380                Can you speak English ?  negative  0.997171
381                         Namanya siapa?  negative  0.893093
382                   Mana gua tau selesai  negative  0.997722

[383 rows x 3 columns]

**Sudah didapat hasil klasifikasi dari data teks**

In [ ]:
df_klasifikasi = df3.drop(columns = 'score')
df_klasifikasi

data teks     label
0                             Makasii ????  negative
1                  Sekarang jam berapa ya?  negative
2     Hp gua udah 4g gausah diaktifin lagi  negative
3                                 Semlekom  negative
4    Sejag kapan nama saiia blue (Har Har)   neutral
..                                     ...       ...
378      Punya anjir. Bacod mulu dari tadi  negative
379          Terima kasih atas infonya bak  positive
380                Can you speak English ?  negative
381                         Namanya siapa?  negative
382                   Mana gua tau selesai  negative

[383 rows x 2 columns]

# __2. Preprocessing data__

Menggunakan proses preprocessing data seperti berikut ini:
1. Case Folding
2. Filtering
3. Stopword
4. Stemming

Menginstal library sastrawi untuk preprocessing NLP bahasa Indonesia

In [ ]:
!pip -q install sastrawi

     |████████████████████████████████| 209 kB 5.3 MB/s 


In [ ]:
import nltk
nltk.download("stopwords")


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [ ]:
#import library yang dibutuhkan
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline


In [ ]:
df_klasifikasi.head()

data teks     label
0                           Makasii ????  negative
1                Sekarang jam berapa ya?  negative
2   Hp gua udah 4g gausah diaktifin lagi  negative
3                               Semlekom  negative
4  Sejag kapan nama saiia blue (Har Har)   neutral

In [ ]:
df_klasifikasi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 383 entries, 0 to 382
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   data teks  383 non-null    object
 1   label      383 non-null    object
dtypes: object(2)
memory usage: 6.1+ KB


## __2.1 Case Folding__

Case Folding digunakan untuk bisa memgubah kata/kalimat menjadi lowecase, menghapus angka dan tanda baca, menghapus whitespace (karakter kosong) yang mana tidak perlukan dalam pemodelan NLP

In [ ]:
# Import library regex
import re

# Buat fungction langkah dari case folding dan filtering
def caseFolding(text):
  text = text.lower() # Mengubah teks menjadi lower case
  text = re.sub(r'https?://\S+|www\.\S+', '', text) # Menghapus URL
  text = re.sub(r'[0-9]+', '', text) # Menghapus angka 
  text = re.sub(r'[^\w\s]', '', text) # Menghapus karakter tanda baca
  text = text.strip()
  return text

In [ ]:
raw_sample = df_klasifikasi['data teks'].iloc[381]
case_folding = caseFolding(raw_sample)

In [ ]:
print('Raw data \t:', raw_sample)
print('Case folding \t', case_folding)

Raw data 	: Namanya siapa?
Case folding 	 namanya siapa


## __2.2 Word Normalization__

Word Normalization digunakan untuk mengubah kata/kalimat slang menjadi seperti kata bakunya.

In [ ]:
# Download corpus slang word

!wget https://raw.githubusercontent.com/ksnugroho/klasifikasi-spam-sms/master/data/key_norm.csv

--2022-07-12 06:11:21--  https://raw.githubusercontent.com/ksnugroho/klasifikasi-spam-sms/master/data/key_norm.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.108.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 80969 (79K) [text/plain]
Saving to: ‘key_norm.csv’

key_norm.csv        100%[===================>]  79.07K  --.-KB/s    in 0.01s   

2022-07-12 06:11:21 (5.83 MB/s) - ‘key_norm.csv’ saved [80969/80969]



In [ ]:
# Meload data key_norm.csv
key_norm = pd.read_csv('/content/key_norm.csv')

def text_normalize(text):
  text = ' '.join([key_norm[key_norm['singkat']==word]['hasil'].values[0] if (key_norm['singkat']==word).any() else word for word in text.split()])
  text = str.lower(text)
  return text


In [ ]:
raw_sample = df_klasifikasi['data teks'].iloc[381]
case_folding = caseFolding(raw_sample)
normalize = text_normalize(case_folding)

In [ ]:
print('Raw data \t:', raw_sample)
print('Case folding \t', case_folding)
print('text_normalize \t', normalize)

Raw data 	: Namanya siapa?
Case folding 	 namanya siapa
text_normalize 	 namanya siapa


## __2.3 Stopword Removal__

Stopword Removal (Filtering) adalah tahap mengambil kata-kata penting dari hasil token dengan menggunakan algoritma stoplist (membuang kata kurang penting) atau wordlist (menyimpan kata penting).

In [ ]:
# Import library stopword

from nltk.corpus import stopwords

stopwords_ind = stopwords.words('indonesian')

In [ ]:
len(stopwords_ind)

758

Stopword Indonesia

In [ ]:
# Membuat fungsi stopword removal

stopwords_ind = stopwords_ind

def remove_stop_word(text):
  clean_word = []
  text = text.split()
  for word in text:
    if word not in stopwords_ind:
      clean_word.append(word)
  return " ".join(clean_word)


In [ ]:
raw_sample = df_klasifikasi['data teks'].iloc[381]
case_folding = caseFolding(raw_sample)
normalize = text_normalize(case_folding)
stopword_removal = remove_stop_word(normalize)

In [ ]:
print('Raw data \t:', raw_sample)
print('Case folding \t', case_folding)
print('text_normalize \t', normalize)
print('Stop word \t', stopword_removal)

Raw data 	: Namanya siapa?
Case folding 	 namanya siapa
text_normalize 	 namanya siapa
Stop word 	 namanya


## __2.4 Stemming__

Stemming adalah proses Mengubah kata ke bentuk dasarnya, namun bentuk dasar tersebut tidak berarti sama dengan akar kata (root word).

In [ ]:
from nltk.text import TextCollection
from Sastrawi.Stemmer.StemmerFactory import StemmerFactory

factory = StemmerFactory()
stemmer = factory.create_stemmer()

#Di sini kita buat fungsi stemming

def stemming(text):
  text = stemmer.stem(text)
  return text


In [ ]:
raw_sample = df_klasifikasi['data teks'].iloc[381]
case_folding = caseFolding(raw_sample)
normalize = text_normalize(case_folding)
stopword_removal = remove_stop_word(normalize)
text_stemming = stemming(stopword_removal)


In [ ]:
print ('Raw data \t:',raw_sample)
print ('Case folding \t:', case_folding)
print ('text_normalize \t:', normalize)
print ('Stop word \t:', stopword_removal)
print ('Stemming \t:', text_stemming)


Raw data 	: Namanya siapa?
Case folding 	: namanya siapa
text_normalize 	: namanya siapa
Stop word 	: namanya
Stemming 	: nama


## __2.5 Pre-processing Pipeline__

In [ ]:
# Disini kita buat pipelinenya (menggabungkan semua fungsi di atas)

def text_preprocessing(text):
  text = caseFolding(text)
  text = text_normalize(text)
  text = remove_stop_word(text)
  text = stemming(text)
  return text


In [ ]:
#Time untuk mendapatkan waktu komputasi
%%time 
#proses menjalankan fungsi text_preprocessing pada data teks
df_klasifikasi['clean data teks'] = df_klasifikasi['data teks'].apply(text_preprocessing)


CPU times: user 36.8 s, sys: 153 ms, total: 36.9 s
Wall time: 45.6 s


In [ ]:
#Simpan data , jika anda ingin menggunakan data clean_teks
df_klasifikasi.to_csv('clean data teks.csv')

In [ ]:
df_klasifikasi.tail()

data teks     label             clean data teks
378  Punya anjir. Bacod mulu dari tadi  negative            anjir bacod mulu
379      Terima kasih atas infonya bak  positive  terima kasih informasi bak
380            Can you speak English ?  negative               speak english
381                     Namanya siapa?  negative                        nama
382               Mana gua tau selesai  negative                 tau selesai

# __3. Feature Engineering / Feature Extraction__

In [ ]:
# Pisahkan kolom feature dan target

X = df_klasifikasi['clean data teks']
y = df_klasifikasi['label']

In [ ]:
# Feature pada variabel X
X

0                            makasii
1                             jam ya
2                       hp diaktifin
3                           semlekom
4      sejag nama saiia blue har har
                   ...              
378                 anjir bacod mulu
379       terima kasih informasi bak
380                    speak english
381                             nama
382                      tau selesai
Name: clean data teks, Length: 383, dtype: object

In [ ]:
# target pada variabel y
y

0      negative
1      negative
2      negative
3      negative
4       neutral
         ...   
378    negative
379    positive
380    negative
381    negative
382    negative
Name: label, Length: 383, dtype: object

## __3.1 Feature Extraction (TF-ID)__

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

#Unigram
tf_idf = TfidfVectorizer(ngram_range=(1,1))
tf_idf.fit(X)

X_tf_idf = tf_idf.transform(X)


In [ ]:
#Menampilkan vocabulary dari TF-IDF
tf_idf.vocabulary_

{'ach': 0,
 'aduh': 1,
 'aduuhh': 2,
 'ae': 3,
 'ah': 4,
 'ahh': 5,
 'alaah': 6,
 'alhamdullilah': 7,
 'allaikum': 8,
 'am': 9,
 'aman': 10,
 'anjay': 11,
 'anjeng': 12,
 'anjir': 13,
 'anjirrr': 14,
 'anying': 15,
 'apa': 16,
 'apaansihgt': 17,
 'apasih': 18,
 'arigato': 19,
 'assalamu': 20,
 'assalamualaiku': 21,
 'assalamualaikumin': 22,
 'asslamualaikum': 23,
 'auk': 24,
 'auw': 25,
 'ayam': 26,
 'bacod': 27,
 'bae': 28,
 'bagus': 29,
 'bahasa': 30,
 'bak': 31,
 'balas': 32,
 'bang': 33,
 'banget': 34,
 'bantu': 35,
 'barusan': 36,
 'beli': 37,
 'belit': 38,
 'bener': 39,
 'bentar': 40,
 'berhentiin': 41,
 'besok': 42,
 'bgni': 43,
 'bgus': 44,
 'bicara': 45,
 'bikin': 46,
 'bilang': 47,
 'bingung': 48,
 'blue': 49,
 'bngt': 50,
 'bnykn': 51,
 'bobo': 52,
 'bodoamayttt': 53,
 'bodoh': 54,
 'bom': 55,
 'book': 56,
 'boot': 57,
 'bosan': 58,
 'bosen': 59,
 'bot': 60,
 'break': 61,
 'butuh': 62,
 'bye': 63,
 'cai': 64,
 'cantik': 65,
 'capek': 66,
 'cek': 67,
 'cepat': 68,
 'chat': 69

In [ ]:
#Melihat Jumlah Fitur
print (len(tf_idf.get_feature_names_out()))

314


In [ ]:
# Melihat vektor kata

X_tf_idf = tf_idf.transform(X).toarray()
X_tf_idf



array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [ ]:
data_tf_idf = pd.DataFrame(X_tf_idf,columns=tf_idf.get_feature_names_out())
data_tf_idf


ach  aduh  aduuhh   ae   ah  ahh  alaah  alhamdullilah  allaikum   am  \
0    0.0   0.0     0.0  0.0  0.0  0.0    0.0            0.0       0.0  0.0   
1    0.0   0.0     0.0  0.0  0.0  0.0    0.0            0.0       0.0  0.0   
2    0.0   0.0     0.0  0.0  0.0  0.0    0.0            0.0       0.0  0.0   
3    0.0   0.0     0.0  0.0  0.0  0.0    0.0            0.0       0.0  0.0   
4    0.0   0.0     0.0  0.0  0.0  0.0    0.0            0.0       0.0  0.0   
..   ...   ...     ...  ...  ...  ...    ...            ...       ...  ...   
378  0.0   0.0     0.0  0.0  0.0  0.0    0.0            0.0       0.0  0.0   
379  0.0   0.0     0.0  0.0  0.0  0.0    0.0            0.0       0.0  0.0   
380  0.0   0.0     0.0  0.0  0.0  0.0    0.0            0.0       0.0  0.0   
381  0.0   0.0     0.0  0.0  0.0  0.0    0.0            0.0       0.0  0.0   
382  0.0   0.0     0.0  0.0  0.0  0.0    0.0            0.0       0.0  0.0   

     ...  wkwkwkwkwkwk  wokee  woy   wr        ya  yaiyalah  yaudahlah  yeay  \
0    ...           0.0    0.0  0.0  0.0  0.000000       0.0        0.0   0.0   
1    ...           0.0    0.0  0.0  0.0  0.575431       0.0        0.0   0.0   
2    ...           0.0    0.0  0.0  0.0  0.000000       0.0        0.0   0.0   
3    ...           0.0    0.0  0.0  0.0  0.000000       0.0        0.0   0.0   
4    ...           0.0    0.0  0.0  0.0  0.000000       0.0        0.0   0.0   
..   ...           ...    ...  ...  ...       ...       ...        ...   ...   
378  ...           0.0    0.0  0.0  0.0  0.000000       0.0        0.0   0.0   
379  ...           0.0    0.0  0.0  0.0  0.000000       0.0        0.0   0.0   
380  ...           0.0    0.0  0.0  0.0  0.000000       0.0        0.0   0.0   
381  ...           0.0    0.0  0.0  0.0  0.000000       0.0        0.0   0.0   
382  ...           0.0    0.0  0.0  0.0  0.000000       0.0        0.0   0.0   

      yh  zzz  
0    0.0  0.0  
1    0.0  0.0  
2    0.0  0.0  
3    0.0  0.0  
4    0.0  0.0  
..   ...  ...  
378  0.0  0.0  
379  0.0  0.0  
380  0.0  0.0  
381  0.0  0.0  
382  0.0  0.0  

[383 rows x 314 columns]

## __3.2 Feature Selection__

In [ ]:
X = np.array(data_tf_idf) #data_unigram
y = np.array(y)

X

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [ ]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2

# Proses Algoritma Chi-Square
chi2_features = SelectKBest(chi2, k=152) #k=152
X_kbest_features = chi2_features.fit_transform(X, y)

# Melihat data
print('Original features:', X.shape[1])
print('Selected features:', X_kbest_features.shape[1])

Original features: 314
Selected features: 152


In [ ]:
# Membuat datframe nilai setiap feature dari chi-square

data_chi2 = pd.DataFrame(chi2_features.scores_, columns=['Nilai'])
data_chi2

Nilai
0    0.565232
1    1.224226
2    0.565232
3    0.501920
4    3.438192
..        ...
309  5.509804
310  0.883245
311  4.044155
312  3.812514
313  2.754902

[314 rows x 1 columns]

In [ ]:
#Menampilkan fitur beserta nilainya
feature = tf_idf.get_feature_names_out()
data_chi2['Fitur'] = feature
data_chi2

Nilai      Fitur
0    0.565232        ach
1    1.224226       aduh
2    0.565232     aduuhh
3    0.501920         ae
4    3.438192         ah
..        ...        ...
309  5.509804   yaiyalah
310  0.883245  yaudahlah
311  4.044155       yeay
312  3.812514         yh
313  2.754902        zzz

[314 rows x 2 columns]

In [ ]:
# Mengurutkan fitur berdasarkan nilai tertinggi
data_chi2.sort_values(by='Nilai', ascending=False)

Nilai        Fitur
280  18.533031  terimakasih
229  17.157895      permisi
143   9.625853        keluh
29    8.776141        bagus
109   8.264706           hi
..         ...          ...
126   0.250960           it
162   0.250960       losing
75    0.250960          cry
288   0.250960          the
211   0.186124          nya

[314 rows x 2 columns]

In [ ]:
#Menampilkan mask pada feature yang terseleksi

#False berarti fitur tidak terpilih dan True berarti fitur terpilih
mask = chi2_features.get_support()
mask


array([False,  True, False, False,  True, False,  True,  True, False,
       False,  True, False,  True,  True, False, False,  True, False,
       False, False, False,  True,  True,  True, False, False, False,
       False, False,  True, False,  True,  True,  True,  True, False,
        True, False, False, False, False, False, False, False, False,
       False, False,  True,  True,  True,  True, False,  True, False,
       False,  True,  True, False, False,  True, False, False,  True,
       False,  True, False, False,  True, False, False,  True, False,
        True, False,  True, False, False, False, False,  True, False,
       False, False, False,  True,  True, False,  True,  True,  True,
        True,  True, False,  True,  True,  True, False, False,  True,
        True,  True, False, False,  True,  True,  True, False, False,
       False,  True, False, False, False,  True, False, False, False,
        True, False, False,  True, False,  True,  True,  True,  True,
       False,  True,

In [ ]:
#Menampilkan fitur-fitur terpilih berdasarkan mask atau nilai tertinggi

new_features = []

for bool, f in zip(mask,feature):
  if bool:
    new_features.append(f)
  selected_features = new_features
selected_features

['aduh',
 'ah',
 'alaah',
 'alhamdullilah',
 'aman',
 'anjeng',
 'anjir',
 'apa',
 'assalamualaiku',
 'assalamualaikumin',
 'asslamualaikum',
 'bagus',
 'bak',
 'balas',
 'bang',
 'banget',
 'barusan',
 'bilang',
 'bingung',
 'blue',
 'bngt',
 'bobo',
 'bom',
 'book',
 'bosen',
 'butuh',
 'cai',
 'cek',
 'cinta',
 'complient',
 'coyyy',
 'deh',
 'diam',
 'donk',
 'dude',
 'dukung',
 'eh',
 'english',
 'erti',
 'fa',
 'full',
 'gai',
 'go',
 'guna',
 'hadeeeeuh',
 'halo',
 'halon',
 'har',
 'hi',
 'hobby',
 'hu',
 'ihhh',
 'informasi',
 'internet',
 'isi',
 'istri',
 'jam',
 'jaring',
 'jawabanya',
 'jjkakak',
 'jumpa',
 'kabar',
 'kasih',
 'keluh',
 'klik',
 'komplain',
 'komputer',
 'ku',
 'kuya',
 'kwkwkwkw',
 'kwlwkwkwkkwkwkw',
 'lahir',
 'layan',
 'logout',
 'loh',
 'love',
 'lumayan',
 'maaf',
 'makasiih',
 'makasiiihh',
 'maksud',
 'malam',
 'mana',
 'masmba',
 'maya',
 'mbak',
 'menyamoaikan',
 'mi',
 'milih',
 'min',
 'misi',
 'muncul',
 'nama',
 'nih',
 'niih',
 'njit',
 'nnya

In [ ]:
# Membuat vocabulary baru berdasarkan fitur yang terseleksi
#Ini digunakan untuk mengenerate fitur vector tf-idf

new_selected_features = {}

for (k,v) in tf_idf.vocabulary_.items():
    if k in selected_features:
        new_selected_features[k] = v

new_selected_features

{'aduh': 1,
 'ah': 4,
 'alaah': 6,
 'alhamdullilah': 7,
 'aman': 10,
 'anjeng': 12,
 'anjir': 13,
 'apa': 16,
 'assalamualaiku': 21,
 'assalamualaikumin': 22,
 'asslamualaikum': 23,
 'bagus': 29,
 'bak': 31,
 'balas': 32,
 'bang': 33,
 'banget': 34,
 'barusan': 36,
 'bilang': 47,
 'bingung': 48,
 'blue': 49,
 'bngt': 50,
 'bobo': 52,
 'bom': 55,
 'book': 56,
 'bosen': 59,
 'butuh': 62,
 'cai': 64,
 'cek': 67,
 'cinta': 70,
 'complient': 72,
 'coyyy': 74,
 'deh': 79,
 'diam': 84,
 'donk': 85,
 'dude': 87,
 'dukung': 88,
 'eh': 89,
 'english': 90,
 'erti': 91,
 'fa': 93,
 'full': 94,
 'gai': 95,
 'go': 98,
 'guna': 99,
 'hadeeeeuh': 100,
 'halo': 103,
 'halon': 104,
 'har': 105,
 'hi': 109,
 'hobby': 113,
 'hu': 117,
 'ihhh': 120,
 'informasi': 122,
 'internet': 123,
 'isi': 124,
 'istri': 125,
 'jam': 127,
 'jaring': 128,
 'jawabanya': 129,
 'jjkakak': 131,
 'jumpa': 133,
 'kabar': 135,
 'kasih': 139,
 'keluh': 143,
 'klik': 146,
 'komplain': 150,
 'komputer': 151,
 'ku': 152,
 'kuya': 

In [ ]:
#Melihat jumlah fitur baru 
#Ini sesuai dengan nilai k pada chi-square
len(new_selected_features)

152

In [ ]:
#Memilih fitur terbaik dan memasukannya kedalam Dataframe
data_selected_features = pd.DataFrame(X_kbest_features,columns=selected_features)
data_selected_features

aduh   ah  alaah  alhamdullilah  aman  anjeng     anjir  apa  \
0     0.0  0.0    0.0            0.0   0.0     0.0  0.000000  0.0   
1     0.0  0.0    0.0            0.0   0.0     0.0  0.000000  0.0   
2     0.0  0.0    0.0            0.0   0.0     0.0  0.000000  0.0   
3     0.0  0.0    0.0            0.0   0.0     0.0  0.000000  0.0   
4     0.0  0.0    0.0            0.0   0.0     0.0  0.000000  0.0   
..    ...  ...    ...            ...   ...     ...       ...  ...   
378   0.0  0.0    0.0            0.0   0.0     0.0  0.551591  0.0   
379   0.0  0.0    0.0            0.0   0.0     0.0  0.000000  0.0   
380   0.0  0.0    0.0            0.0   0.0     0.0  0.000000  0.0   
381   0.0  0.0    0.0            0.0   0.0     0.0  0.000000  0.0   
382   0.0  0.0    0.0            0.0   0.0     0.0  0.000000  0.0   

     assalamualaiku  assalamualaikumin  ...  waah   wb  wkwkwkwkwkwk  wokee  \
0               0.0                0.0  ...   0.0  0.0           0.0    0.0   
1               0.0                0.0  ...   0.0  0.0           0.0    0.0   
2               0.0                0.0  ...   0.0  0.0           0.0    0.0   
3               0.0                0.0  ...   0.0  0.0           0.0    0.0   
4               0.0                0.0  ...   0.0  0.0           0.0    0.0   
..              ...                ...  ...   ...  ...           ...    ...   
378             0.0                0.0  ...   0.0  0.0           0.0    0.0   
379             0.0                0.0  ...   0.0  0.0           0.0    0.0   
380             0.0                0.0  ...   0.0  0.0           0.0    0.0   
381             0.0                0.0  ...   0.0  0.0           0.0    0.0   
382             0.0                0.0  ...   0.0  0.0           0.0    0.0   

      wr        ya  yaiyalah  yeay   yh  zzz  
0    0.0  0.000000       0.0   0.0  0.0  0.0  
1    0.0  0.575431       0.0   0.0  0.0  0.0  
2    0.0  0.000000       0.0   0.0  0.0  0.0  
3    0.0  0.000000       0.0   0.0  0.0  0.0  
4    0.0  0.000000       0.0   0.0  0.0  0.0  
..   ...       ...       ...   ...  ...  ...  
378  0.0  0.000000       0.0   0.0  0.0  0.0  
379  0.0  0.000000       0.0   0.0  0.0  0.0  
380  0.0  0.000000       0.0   0.0  0.0  0.0  
381  0.0  0.000000       0.0   0.0  0.0  0.0  
382  0.0  0.000000       0.0   0.0  0.0  0.0  

[383 rows x 152 columns]

# __4. Modelling (Machine Learning)__

In [ ]:
#Import Library
from sklearn.model_selection import train_test_split

#Algoritme

#Algoritme Support Vector Machine
from sklearn import svm
#Naive Bayes
from sklearn.naive_bayes import MultinomialNB

#Algoritma Random Forest
from sklearn.ensemble import RandomForestClassifier

#Algoritma Decision Tree
from sklearn.tree import DecisionTreeClassifier

In [ ]:
x = data_selected_features
y = df_klasifikasi['label']
#Memisihkan data training dan data testing dengan perbandingan 75:25
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.25, random_state=6)

In [ ]:
#print
print('Banyak data x_train :',len(x_train))
print('Banyak data x_test  :',len(x_test))
print('Banyak data y_train :',len(y_train))
print('Banyak data y_test  :',len(y_test))

Banyak data x_train : 287
Banyak data x_test  : 96
Banyak data y_train : 287
Banyak data y_test  : 96


In [ ]:
#Training Model

from datetime import datetime
start_time = datetime.now()

#algoritme fitting

#text_algorithm = MultinomialNB()
text_algorithm = svm.SVC(kernel="rbf", C=1.0)
#text_algorithm = RandomForestClassifier()
#text_algorithm = DecisionTreeClassifier()
model = text_algorithm.fit(x_train, y_train)


end_time = datetime.now()
result_time  =end_time-start_time
print("Duration:",result_time)


Duration: 0:00:00.020116


In [ ]:
#Prediksi
predicted = model.predict(x_test)
predicted

array(['negative', 'negative', 'negative', 'negative', 'negative',
       'negative', 'positive', 'negative', 'positive', 'negative',
       'negative', 'negative', 'negative', 'positive', 'positive',
       'negative', 'negative', 'negative', 'negative', 'negative',
       'negative', 'negative', 'positive', 'negative', 'negative',
       'positive', 'positive', 'positive', 'negative', 'negative',
       'positive', 'positive', 'negative', 'positive', 'negative',
       'positive', 'negative', 'positive', 'negative', 'negative',
       'positive', 'negative', 'negative', 'positive', 'positive',
       'positive', 'negative', 'negative', 'positive', 'negative',
       'positive', 'negative', 'negative', 'negative', 'negative',
       'positive', 'negative', 'negative', 'positive', 'negative',
       'positive', 'positive', 'positive', 'negative', 'negative',
       'positive', 'negative', 'negative', 'negative', 'positive',
       'positive', 'negative', 'negative', 'negative', 'positi

# __5. Model Evaluation__

In [ ]:
#Library evaluasi
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

#Mengkomputasi performa model
NewprediksiBenar = (predicted == y_test).sum()
NewprediksiSalah = (predicted != y_test).sum()
    
print("prediksi benar: ", NewprediksiBenar, " data")
print("prediksi salah: ", NewprediksiSalah, " data")
print("Akurasi Algoritma: ", NewprediksiBenar/(NewprediksiBenar+NewprediksiSalah)*100,"%")

# Menggunakan Confusion Matrix untuk Model Evaluasi
CM = confusion_matrix(y_test,predicted)

TN = CM[0][0]
FN = CM[1][0]
TP = CM[1][1]
FP = CM[0][1]
precision    = TP/(TP+FP)
recall       = TP/(TP+FN)
f1 = 2*recall*precision/(recall+precision)

print ("TRUE NEGATIVE (TN):",TN)
print ("FALSE NEGATIVE (FN):",FN)
print ("TRUE POSITIVE (TP):",TP)
print ("FALSE POSITIVE (FP):",FP)
print ("PRECISION:",precision*100,"%")
print ("RECALL:",recall*100,"%")
print ("F1:",f1*100,"%")

print(classification_report(y_test,predicted))

prediksi benar:  75  data
prediksi salah:  21  data
Akurasi Algoritma:  78.125 %
TRUE NEGATIVE (TN): 52
FALSE NEGATIVE (FN): 5
TRUE POSITIVE (TP): 2
FALSE POSITIVE (FP): 0
PRECISION: 100.0 %
RECALL: 28.57142857142857 %
F1: 44.44444444444445 %
              precision    recall  f1-score   support

    negative       0.85      0.96      0.90        54
     neutral       1.00      0.12      0.21        17
    positive       0.64      0.84      0.72        25

    accuracy                           0.78        96
   macro avg       0.83      0.64      0.61        96
weighted avg       0.82      0.78      0.73        96



In [ ]:
#Menggunakan Cross Validation untuk memvalidasi data

from sklearn.model_selection import ShuffleSplit
from sklearn.model_selection import cross_val_score

cv        = ShuffleSplit(n_splits=10, test_size=0.25, random_state=6)
accuracy  = (cross_val_score(model,x,y,cv=cv,scoring='accuracy'))
score_accuracy = np.mean(accuracy)
print ("accuracy: ",accuracy)
print ("accuracy: ",score_accuracy)

accuracy:  [0.78125    0.76041667 0.82291667 0.75       0.80208333 0.79166667
 0.75       0.76041667 0.76041667 0.80208333]
accuracy:  0.778125


# Kesimpulan

Setelah dilakukannya pemodelan dengan berbagai macam model klasifikasi seperti Random Forest, Decision Tree, Naive Bayes, Support Vector Machine, dll. 

Didapatkan hasil algoritma model yang paling tinggi akurasinya yaitu menggunakan SVM (Support Vector Machine) dengan akurasi tertinggi 78.1% dengan evaluasi model confusion matrix untuk Precission = 100%, Recall = 28.57, F1 = 44.44%. Serta untuk Cross Validation mendapatkan score 77%.
